lib/services/watcher_test.go

```diff
@@ -224,8 +228,136 @@ func TestLockWatcher(t *testing.T) {
 		t.Fatal("Lock watcher subscription has unexpectedly exited.")
 	case <-time.After(2 * time.Second):
 	}
-	require.Nil(t, w.FindLockInForce(target))
-	require.Empty(t, resourceDiff(w.FindLockInForce(target2), lock2))
+	require.NoError(t, w.CheckLockInForce(constants.LockingModeBestEffort, target))
+	expectLockInForce(t, lock2, w.CheckLockInForce(constants.LockingModeBestEffort, target2))
+}
+
+func TestLockWatcherStale(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	clock := clockwork.NewFakeClock()
+
+	bk, err := lite.NewWithConfig(ctx, lite.Config{
+		Path:             t.TempDir(),
+		PollStreamPeriod: 200 * time.Millisecond,
+		Clock:            clock,
+	})
+	require.NoError(t, err)
+
+	type client struct {
+		services.Access
+		types.Events
+	}
+
+	access := local.NewAccessService(bk)
+	events := &withUnreliability{Events: local.NewEventsService(bk, nil)}
+	w, err := services.NewLockWatcher(ctx, services.LockWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component:   "test",
+			RetryPeriod: 200 * time.Millisecond,
+			Client: &client{
+				Access: access,
+				Events: events,
+			},
+			Clock: clock,
+		},
+	})
+	require.NoError(t, err)
+	t.Cleanup(w.Close)
+
+	// Subscribe to lock watcher updates.
+	target := types.LockTarget{Node: "node"}
+	require.NoError(t, w.CheckLockInForce(constants.LockingModeBestEffort, target))
+	sub, err := w.Subscribe(ctx, target)
+	require.NoError(t, err)
+	t.Cleanup(func() { require.NoError(t, sub.Close()) })
+
+	// Close the underlying watcher. Until LockMaxStaleness is exceeded, no error
+	// should be returned.
+	events.setUnreliable(true)
+	bk.CloseWatchers()
+	select {
+	case event := <-sub.Events():
+		t.Fatalf("Unexpected event: %v.", event)
+	case <-sub.Done():
+		t.Fatal("Lock watcher subscription has unexpectedly exited.")
+	case <-time.After(2 * time.Second):
```

Comment:
This is supposed to test an event did *not* occur, the duration was a guess to make the test wait long enough for any unexpected events.

Reply:
Tests that sleep on the happy path are a pet peeve of mine, since I don't want to wait any time more than the necessary. We could probably get away with a smaller number here, at least, but up to you - this isn't a big deal.